src/UserSpaceInstrumentation/AllocateInTracee.cpp

```diff
@@ -62,6 +60,21 @@ namespace {
     return ErrorMessage(absl::StrFormat("Failed to read from tracee's memory: %s",
                                         backup_or_error.error().message()));
   }
+  const std::vector<uint8_t> backup = std::move(backup_or_error.value());
+
+  orbit_base::unique_resource restore_memory_and_registers_on_return{
+      pid, [start_address, backup, &original_registers](pid_t pid) {
+        auto restore_memory_result = WriteTraceesMemory(pid, start_address, backup);
+        if (restore_memory_result.has_error()) {
+          ORBIT_FATAL("Unable to restore memory state of tracee: %s",
+                      restore_memory_result.error().message());
+        }
+        auto restore_registers_result = original_registers.RestoreRegisters();
+        if (restore_registers_result.has_error()) {
+          ORBIT_FATAL("Unable to restore register state of tracee: %s",
+                      restore_registers_result.error().message());
+        }
```

Comment:
This is now done in the case of all the errors (that's the point of the change of course). While conceptually correct, I wonder if it could trigger the `ORBIT_FATAL` in some of the error cases. Not sure how to check that though.

Reply:
I'm also wondering if we should always to both restores, for example `original_registers.RestoreRegisters()` even if `WriteTraceesMemory(pid, start_address, std::vector<uint8_t>{0x0f, 0x05});` failed right away.

So again I guess I'd like you hear your thoughts.

Reply:
I think it is fine as is. Restoring anything after a failed `WriteTraceesMemory(pid, start_address, std::vector<uint8_t>{0x0f, 0x05});` is probably not needed and restoring the registers is only required after the first `RestoreRegisters`. For the sake of simplicity I'd keep it here. 

I think using `ORBIT_FATAL` is fine here: If we fail in the middle of modifing the state of the tracee and can't restore it we can't really continue in a meaningful way. Also: I can't remember ever seeing any of the errors potentially produced in this function.

Reply:
So, for the series "ask a question and then do the opposite", I split the restoration code and placed it in the "right" places. Might as well, it cost me nothing.